sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -361,7 +362,8 @@ class HDBSCAN(ClusterMixin, BaseEstimator):
         Outliers are labeled as follows:
         - Noisy samples are given the label -1.
         - Samples with infinite elements (+/- np.inf) are given the label -2.
-        - Samples with missing data are given the label -3.
+        - Samples with missing data are given the label -3, even if they
+          also have infinite elements.
```

Comment:
I slightly prefer `inf` having a higher priority when both are in a sample. In other words, labeling it as `-2`.

How did you decide on `-3` as the default?

Reply:
Going with the semantics that @lmcinnes had discussed with me regarding how HDBSCAN is used as a downstream for umap, the presence of `np.nan` indicates disconnection from the target manifold, or in alternative workflows data that is missing. Note this is all conditional on my understanding, so @lmcinnes please feel free to correct me if I am wrong.

For the prior, even if the point has an `np.inf`, its connection to the target manifold is sensitive to the presence of `np.nan` whereas a disconnected point is indifferent to the presence of `np.inf`. In this sense the `np.nan` conveys more information (or at least with a higher precendence) hence my preference for it to dominate.

For the latter case, the presence of `np.inf` conveys valid and usable information, whereas the presence of `np.nan` indicates invalidity of the sample. Inference can be made on samples with `np.inf` but not necessarily with `np.nan`, hence my preference for `np.nan` to dominate.